### PR TITLE
add error diagnostics for E0437, E0438

### PR DESCRIPTION
**Pull request by apasel422**

### CODE REVIEW

File: src/librustc_resolve/diagnostics.rs [hunk 1]
```diff
@@ -272,6 +272,81 @@ See the 'Use Declarations' section of the reference for more information
 on this topic:
 
 http://doc.rust-lang.org/reference.html#use-declarations
+"##,
+
+E0407: r##"
+Trait impls can only declare methods that are members of the trait in question.
```
Comment (tamird) [inline src/librustc_resolve/diagnostics.rs]: nit: declare isn't quite right here. the trait itself "declares" the methods, and the impl implements them.

File: src/librustc_resolve/diagnostics.rs [hunk 2]
```diff
@@ -272,6 +272,81 @@ See the 'Use Declarations' section of the reference for more information
...
+
+```
+trait Foo {}
+
+impl Foo for i32 {}
+```
+"##,
+
+E0437: r##"
+Trait impls can only declare associated types that are members of the trait in
```
Comment (tamird) [inline src/librustc_resolve/diagnostics.rs]: ditto

File: src/librustc_resolve/diagnostics.rs (hunk 1 above)
Comment (apasel422) [inline src/librustc_resolve/diagnostics.rs]: Should "implement" be used for each of the item types (not just methods)?

**Comment by apasel422** [comment]
Nits addressed.

File: src/librustc_resolve/diagnostics.rs [hunk 3]
```diff
@@ -272,6 +272,81 @@ See the 'Use Declarations' section of the reference for more information
...
+
+E0407: r##"
+Trait impls can only implement methods that are members of the trait in
+question. This error indicates that you attempted to implement a method whose
+name does not match the name of any method in the trait.
+
+Here is an example that demonstrates the error:
+
+```
+trait Foo {}
```
Comment (Manishearth) [inline src/librustc_resolve/diagnostics.rs]: add a trait method in this example

File: src/librustc_resolve/diagnostics.rs [hunk 4]
```diff
@@ -272,6 +272,81 @@ See the 'Use Declarations' section of the reference for more information
...
+
+impl Foo for i32 {
+    fn bar() {}
+}
+```
+
+The solution to this problem is to remove the extraneous method:
```
Comment (Manishearth) [inline src/librustc_resolve/diagnostics.rs]: ... or to implement it in a separate `impl` block

**Comment by apasel422** [comment]
@Manishearth Updated.

**Comment by apasel422** [comment]
Rebased.

**Comment by Manishearth** [comment]
r=me with a trait method added

**Comment by apasel422** [comment]
To which one? This PR no longer addresses E0407.

**Comment by Manishearth** [comment]
@bors r+ rollup

oh, okay

**Pull request merged**